iac-common/src/test/java/org/sonar/iac/common/extension/AnalyzerTest.java

```diff
@@ -0,0 +1,204 @@
+    assertThat(debugLogs.get(0)).isEqualTo("Custom runtime exception");
+    assertThat(debugLogs.get(1)).startsWith("org.sonar.iac.common.extension.ParseException: Cannot parse 'FileWithContent'");
+  }
+
+  @Test
+  void shouldReportErrorOnVisitorScanException() throws IOException {
+    TreeVisitor<InputFileContext> visitorFail = mock(TreeVisitor.class);
+    doThrow(new RuntimeException("Exception when scan mock"))
+      .when(visitorFail).scan(any(InputFileContext.class), any(Tree.class));
+    Analyzer analyzer = new Analyzer("iac", parser, List.of(visitorFail), durationStatistics);
+    List<InputFile> files = List.of(file("Some content"));
+    analyzer.analyseFiles(context, files, progressReport);
+
+    assertThat(logTester.logs(Level.ERROR)).containsExactly("Cannot analyse 'FileWithContent': Exception when scan mock");
+  }
+
+  @Test
+  void shouldReportIllegalStateExceptionWhenFailFastIsTrue() throws IOException {
+    TreeVisitor<InputFileContext> visitorFail = mock(TreeVisitor.class);
+    doThrow(new RuntimeException("Exception when scan mock"))
+      .when(visitorFail).scan(any(InputFileContext.class), any(Tree.class));
+    Analyzer analyzer = new Analyzer("iac", parser, List.of(visitorFail), durationStatistics);
+    List<InputFile> files = List.of(file("Some content"));
+    MapSettings settings = new MapSettings();
+    settings.setProperty(IacSensor.FAIL_FAST_PROPERTY_NAME, true);
+    context.setSettings(settings);
+
+    assertThatThrownBy(() -> analyzer.analyseFiles(context, files, progressReport))
+      .isInstanceOf(IllegalStateException.class)
+      .hasMessage("Exception when analyzing 'FileWithContent'");
+    assertThat(logTester.logs(Level.ERROR)).containsExactly("Cannot analyse 'FileWithContent': Exception when scan mock");
+  }
+
+  @Test
+  void shouldParseAndVisitWithSuccess() throws IOException {
+    TreeVisitor<InputFileContext> visitor = mock(TreeVisitor.class);
+    Analyzer analyzer = new Analyzer("iac", parser, List.of(visitor), durationStatistics);
+    List<InputFile> files = List.of(file("Some content"));
+
+    assertThat(analyzer.analyseFiles(context, files, progressReport)).isTrue();
+  }
+
+  SensorContext sensorContext(boolean cancelled) {
+    SensorContext mock = mock(SensorContext.class);
+    when(mock.isCancelled()).thenReturn(cancelled);
+    var analysisError = mock(DefaultAnalysisError.class);
+    when(analysisError.message(anyString())).thenReturn(analysisError);
+    when(mock.newAnalysisError()).thenReturn(analysisError);
+    return mock;
+  }
+
+  InputFile file(String content) throws IOException {
+    InputFile file = mock(InputFile.class);
+    when(file.contents()).thenReturn(content);
+    when(file.toString()).thenReturn("FileWithContent");
+    return file;
+  }
+
+  InputFile emptyFile() throws IOException {
+    InputFile file = mock(InputFile.class);
+    when(file.contents()).thenReturn("  ");
+    when(file.toString()).thenReturn("EmptyFile");
+    return file;
+  }
```

Comment:
Why are you Mocking everything? Is it not possible to use the SensorContextTester or actual InputFile via `IacTestUtils`?

Reply:
We will try tomorrow to improve it